### ISSUE TITLE
Tell user app has been minimised to tray icon when close button is clicked

### ISSUE THREAD

**Issue by laurent22** [laurent22/joplin#1623]
### Operating system

- Windows
- macOS
- Linux

### Application

- Desktop

When user clicks on the close button of the window, it is in fact minimised to the system tray. This confuses some users who think the app is actually closed when it's not.

To avoid this:

- If `showTrayIcon` is enabled, display a message box when the user click on the Close button. The message should say "Joplin is going to be minimised to its tray icon and will still be running. You may change this behaviour in the Configuration screen." with two buttons: "OK" / "Don't remind me".
- If "OK" is clicked, do nothing.
- If "Don't remind me" is clicked, save the result to a models/Setting variable named "showTrayIconMinimisedMessage" (defaults to `true`) and set it to `false`.
- Later only display this message box, if "showTrayIconMinimisedMessage" is `true`.

**Comment by tessus** [comment]
Yes, please make this configurable, as stated in point 3. I couldn't stand clicking `ok` every time I close the window, which I do instead of hiding it.

**Comment by ratchek** [comment]
I'd like to try and contribute. This is gonna be my first issue. Could you perhaps point me towards the file which should be modified?

**Comment by laurent22** [comment]
@ratchek, the first step will be to build the app using the instructions in BUILD.md. 

Then you'll want to modify `ElectronClient/app/ElectronAppWrapper.js` line 98 and line 103 to display a message box when the user try to close the app and the tray icon is enabled. To display the message you can use `bridge().showInfoMessageBox`.

To add the new setting value, you can edit `models/Setting.js` and add your variable there (of type TYPE_INT). You'll then need to display the message box or not depending on this value.

There are probably a few other details to get right but if you need more info feel free to ask here.

**Comment by ratchek** [comment]
Thanks for the explanation!

**Comment by ratchek** [comment]
Ok, two questions (probably very elementary, sorry for that):
1. Is there any way to see the result of my code aside from running 'yarn dist' over and over again? Some way to recompile just the file I modified and have the changes show up when I run the app?
2. Is there any place where I can read about what different files and functions do? (This isn't strictly necessary for this modification, just curious to get a feel of how the project works)

**Comment by tessus** [comment]
I can only answer your first question:

You can use the script `./run.sh` in the directory `ElectronClient`. This will only start and run the client (with your local changes) without creating a full package.

**Comment by ratchek** [comment]
Thank you @tessus !

**Comment by ratchek** [comment]
Ok. I'm officially stuck. 
I'm having a problem with returning which button was clicked in the dialog box. Googling yielded that I should use the `dialog.showMessageBox` method with a callback function, but non of the functions in `bridge.js` allow me to return the result of that callback to `ElectronAppWrapper.js`. 
One option I see would be to create a dedicated function in 'bridge.js', but I just wanted to make sure it's ok if I go mucking around in that file. Are there perhaps any other ways I should go about this that I'm just not seeing?

**Comment by laurent22** [comment]
You could call bridge().showConfirmMessageBox() with options `{ buttons: [_('OK'), _('Don't remind me')] }`. This function will then return `true` if the first button has been clicked. So it will be true if OK is clicked, and false if "Don't remind me" is clicked.

It's an async function, so you need to use it with `await`.

**Comment by laurent22** [comment]
> Is there any place where I can read about what different files and functions do? (This isn't strictly necessary for this modification, just curious to get a feel of how the project works)

Nothing like this at the moment. All the available documentation is in the "readme" folder but there's not much related to the actual code.

**Comment by tessus** [comment]
Something just came to mind: what about the menu items `Close Window` and `Hide Joplin`. Will they trigger that as well? In my opinion the diaglog box should only appear when the window is closed by clicking on the close window button (the red one on macOS, the X on Wiindows).

**Comment by ratchek** [comment]
Hi guys, 
I've run into a weird issue. Every time I try to run the app or even build it from the ground up, the changes I made to `ElectronClient/app/lib/Setting.js` are erased. All the other modified files stay modified, it's just one line that keeps disappearing from that file specifically. Any ideas what might be happening?

**Comment by tessus** [comment]
You have to change the `Settings.js` file in the directory `ReactNativeClient/lib/models`.

**Comment by tessus** [comment]
The build script copies the lib directory into the ElectronClient source tree, thus changes you make in `ElectronClient/app/lib` will be overwritten.

**Comment by laurent22** [comment]
Yes the issue with the "lib" directory is explained at the top BUILD.md

**Comment by ratchek** [comment]
Got it! Thanks guys.

**Comment by ratchek** [comment]
More issues: 

1. @laurent22, you mentioned that I need to use `await` with `bridge().showConfirmMessageBox()` but it doesn't seem like that's an async function and when I try to put `await` before it, I get a message that await should only be used with async functions.
Disclaimer: I've never used async and await before, and while I did read up on the documentation, and I went through a couple of quick tutorials and examples I found online, it might be that I'm just doing something embarrassingly wrong.

2. @tessus which menu items do you mean? Only ones I could find were the 'File -> Quit' in the top menu and 'Exit' in the tray menu. Neither seems to trigger the popup at the moment. If you meant a different menu item, point me towards it and I'll check it out.

3. When I try to close the window, I get this error: 
`Uncaught Exception:
Error: Settings have not been initialized!
    at Function.value (/home/tomek/TOMEK/Projects/open source/joplin/ElectronClient/app/lib/models/Setting.js:450:27)
    at BrowserWindow.win_.on (/home/tomek/TOMEK/Projects/open source/joplin/ElectronClient/app/ElectronAppWrapper.js:106:19)
    at BrowserWindow.emit (events.js:187:15)`
This happens before the popup window even has a chance to appear (the popup window was working before, when it didn't modify the Setting.js). I've tried inspecting the app before this happens, and the `Setting.keys_` exists and contains the new key, however Setting.cache_ doesn't have my key and it's that last fact that seems to be triggerring the Exception. I tried tracing the code but couldn't understand what exactly pushes the key to the cache. Do I have to somehow trigger `Setting.saveAll`? Oh also, I noticed that the cache only contains 15 keys whereas the Setting object contains more than 30 (which I'm pretty sure you all know and understand why that's happening.

tldr;
I'm not sure which function @laurent22 wanted await'ed and I'm getting a `Settings have not been initialized!` error.

**Comment by tessus** [comment]
> which menu items do you mean?

Don't worry about it for now. The question was rather directed to Laurent, since I didn't see the function call in this discussion.
On macOS there's a `Hide Joplin` menu item in the apps menu `Joplin` and a `Close Window` menu item in the menu `File`.
We can look at it and test it when you create the PR.

**Comment by laurent22** [comment]
1\. You're right it's not async actually, so you can call it directly.

3\. It's not possible to tell without seeing the code. Perhaps create a pull request with what you have?

**Comment by ratchek** [comment]
Done. :)

**Comment by devonzuegel** [comment]
Linking the PR here for future reference: https://github.com/laurent22/joplin/pull/1657 😊
